File: 1498-find-a-corresponding-node-of-a-binary-tree-in-a-clone-of-that-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* res ;
    TreeNode* getTargetCopy(TreeNode* original, TreeNode* cloned, TreeNode* target) {
        Preorder(original,cloned,target);
        return res ;
    }
    
    void Preorder(TreeNode* orig , TreeNode* cloned , TreeNode* target){
        if(orig != NULL){
            if(orig == target) res = cloned ;
            Preorder(orig->left,cloned->left,target);
            Preorder(orig->right,cloned->right,target);
        }
    }
};

File: 1498-find-a-corresponding-node-of-a-binary-tree-in-a-clone-of-that-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public final TreeNode getTargetCopy(final TreeNode o, final TreeNode c, final TreeNode t) {
        return solve(o,c,t);
    }

    public final TreeNode solve(final TreeNode o, final TreeNode c, final TreeNode t){
        if(o == null || c == null) return null ;
        if(o == t) return c ;
        TreeNode x = solve(o.left,c.left,t);
        TreeNode y = solve(o.right,c.right,t);

        if(x != null) return x ;
        if(y != null) return y ;
        return null ;
    }
}
